prover/src/trace/mod.rs

```diff
@@ -38,62 +41,118 @@ mod tests;
 pub trait Trace: Sized {
     /// Base field for this execution trace.
     ///
-    /// All cells of this execution trace contain values which are elements in this filed.
+    /// All cells of this execution trace contain values which are elements in this field.
     type BaseField: StarkField;
 
     // REQUIRED METHODS
     // --------------------------------------------------------------------------------------------
 
-    /// Returns number of columns in this trace.
-    fn width(&self) -> usize;
+    /// Returns a description of how columns of this trace are arranged into trace segments.
+    fn layout(&self) -> &TraceLayout;
 
     /// Returns the number of rows in this trace.
     fn length(&self) -> usize;
 
     /// Returns metadata associated with this trace.
     fn meta(&self) -> &[u8];
 
-    /// Returns value of the cell in the specified column at the specified row of this trace.
-    fn get(&self, col_idx: usize, row_idx: usize) -> Self::BaseField;
+    /// Returns a reference to a [Matrix] describing the main segment of this trace.
+    fn main_segment(&self) -> &Matrix<Self::BaseField>;
+
+    /// Builds and returns the next auxiliary trace segment. If there are no more segments to
+    /// build (i.e., the trace is complete), None is returned.
+    fn build_aux_segment<E>(&mut self, rand_elements: &[E]) -> Option<&Matrix<E>>
```

Comment:
I'm a bit puzzled with the way that `E` is handled. I've tried to do some noodling around this with the Rescue example but I couldn't solve the type parameter related issues popping when constructing a custom `RapTraceTable` and implementing the related methods.
It feels like removing the type parameter for auxiliary segments related methods, and defining instead an additional type `type AuxBaseField: FieldElement<BaseField = Self::BaseField>;` along with the existing `Trace::BaseField` would ease things, though it may require changes on the main Prover flow. What do you think is best? Or have you thought of a workaround that I've missed for implementing the Trace trait on a custom struct?

Reply:
Is the issue caused by the fact that we need to store values of type `E` in the `RapTraceTable`? But `E` is only a type parameter used for methods of `Trace` (rather than being a property of `Trace` itself?

If so, I wonder if adding a generic parameter to `Trace` trait would solve it. Something like `pub trait Trace<E: FieldElement<BaseField = Self::BaseField>>`.

Reply:
I've refactored how auxiliary trace generation works a little in the latest commit. Basically, I removed the need to store auxiliary trace segment in the struct which implements `Trace` trait. This should help, but let me know if you still run into this issue.

Reply:
I'll try to continue but I'm mostly going to be away from the computer this week unfortunately.
Regarding the refactoring, it's great! But I think we may still run into the issue, given that `build_aux_segment<E>()` still requires this type parameter that doesn't come either from the `Trace` trait or the `RapTraceTable` struct. 
I guess it depends on how you wanted the interface for the user defining the auxiliary segments to be. In our approach, we were using some lazy evaluation, storing the provided `init_aux` / `update_aux` (for a single auxiliary segment) in a Box in the RapTraceTable struct (btw there were also some heavy type conflicts with this, so we had gone for [direct specification instead of additional types](https://github.com/Toposware/winterfell/blob/b027c6de528c213d54efb8cadd12200292c3959d/prover/src/trace/trace_table.rs#L83): 
`aux_init: Box<dyn Fn(&[B], &[B], &mut [B])>,`
`aux_update: Box<dyn Fn(usize, &[B], &[B], &mut [B])>,`

Definitely not the cleanest approach, I agree, but if we want `RapTraceTable` to handle multiple segments, such that we don't store the auxiliary segments and we can have some wiggle room for specifying custom RAPs behaviour without having to create additional Structs implementing `Trace`.

Reply:
Yeah - providing `init` and `update` functions would be challenging. This is not a problem if the user is implementing the `Trace` trait directly (at least not as far as I can see), but does cause issues for a "default" implementation of the trait.

I'm not seeing a clean solution to this yet - so, maybe for now we don't try to tackle it?

One possible approach is to move `build_aux_segment()` into the `Prover` trait - but I'm not sure this solves the problem.

Reply:
Yes we can leave the "default" implementation as a To-Do if that is too challenging for now. I would need also more time to think about what could be the best way to handle this.